First cut at a CHERI-UPDATING document. (#448)
CTSRD-CHERI/cheribsd: CHERI-UPDATING.md

# Updating Information for CheriBSD users.

This file contains information about updating CheriBSD.  It supplements
the information in the [UPDATING] file with CheriBSD specific
information.  In particular it explains ABI flag days, our branching structure,

> and our stragety from merging from upstream FreeBSD.

Reviewer: ```suggestion
and our strategy from merging from upstream FreeBSD.
```

## Flag days and major changes

Generally speaking, flag days require a complete rebuild of CheriBSD.
When building with [cheribuild] this is usually handled automatically,
but if necessicary the `--clean` flag can be used.  When building
directly, avoid the `-DNO_CLEAN` make option.

### Flag days on [master]

| Date       | Commit      | Description | Required action |
| ---------- | ----------- | --- | --- |
| 2020-03-06 | [6ce214d1e] | ELF auxargs flags altered | Clean rebuild |

### Flag days on [dev]

| Date       | Commit      | Description | Required action |
| ---------- | ----------- | --- | --- |

> | 2020-03-21 | [7e76d8f71] | C/C++ ABI changes | Update to [LLVM b7f5c847dc] and a clean rebuild.|

Reviewer: How will this appear when we merge to master. Will we give the original date/commit, or that of the merge? We need to be communicating both clearly.

Author: My thought is that we'll give the date of the merge and the merge commit (maybe with the merged commit in the description.)  Awkwardly, that means a double merge where we commit with an XXXX in place of the Commit and then commit the real entry.

Reviewer 2: Or you just push the update to this file separately as a followup change.

| 2020-03-06 | [6ce214d1e] | ELF auxargs flags altered | Clean rebuild |

## Branches

The CheriBSD repository contains a number of branches.  The two main branches
are:

* [dev] - The primary development branch.  Pull requests should generally be
  targeted here.  When using this branch, you should also track the [LLVM dev]
  branch.

* [master] - The default branch, synced periodically with [dev] and kept in
  sync with the [LLVM master] branch.  Outside consumers likely wish to follow
  this branch.

> To aid comparison with upstream FreeBSD we maintain a couple branches of stock

Reviewer: ```suggestion
To aid comparison with upstream FreeBSD we maintain a couple of branches of stock
```

FreeBSD:

* [freebsd-head] - FreeBSD HEAD (from [freebsd/freebsd]) as merged to
  [dev].

> * [freebsd-crossbuild] - FreeBSD HEAD with a roll up of not-yet-merged

Reviewer: Do we rebase this branch or keep merging? If the former we should make sure to clearly highlight that.

Author: We keep merging on the grounds that it's less disruptive to consumers and we're supposed to be working to upstream all of this.

>   cross-building changes required to build on a Mac OS or Linux system.

Reviewer: ```suggestion
  cross-building changes required to build on a macOS or Linux system.
```

> Numerious other branches exist ranging from pull-request branches to long-term

Reviewer: ```suggestion
Numerous other branches exist ranging from pull-request branches to long-term
```

feature development and checkpoints of abandoned work.  We generally delete
pull-request branches after merge.

> ## Merging stragety

Reviewer: ```suggestion
## Merging strategy
```

### Updating [dev]

We typically merge from upstream FreeBSD to [dev] in batches of one week
of changes from the end of Friday UTC.  These are merged one upstream

> commit at a time using [mergify] to aid bisection.  Sometimes we either

Reviewer: ```suggestion
commit at a time using [mergify] to aid bisection.  Sometimes we
```

unless there's a missing `or` half of the sentence.

merge at other times because we need an upstream commit.  In a steady
state we merge weekly, but delay and batch updates if we need extra
platform stability and the [master] branch isn't appropriate.

> Each merge from upstream FreeBSD to [dev] it accompanied by a tag of the form

Reviewer: ```suggestion
Each merge from upstream FreeBSD to [dev] is accompanied by a tag of the form
```

> `freebsd-head-YYYYMMDD`, and merges to [freebsd-head] and [freebsd-crossbuild].

Reviewer: We fast-forward `freebsd-head` so should probably avoid saying it's a merge. Depending on the answer for `freebsd-crossbuild` above, we're also not _merging_.

### Updating [master]

> We typicaly merge to [master] from [dev] at stable points at least a week

Reviewer: ```suggestion
We typically merge to [master] from [dev] at stable points at least a week
```

> apart

Reviewer: ```suggestion
apart.
```

# Updating Information for CheriBSD users.

This file contains information about updating CheriBSD.  It supplements
the information in the [UPDATING] file with CheriBSD specific
information.  In particular it explains ABI flag days, our branching structure,
and our strategy from merging from upstream FreeBSD.

## Flag days and major changes

Generally speaking, flag days require a complete rebuild of CheriBSD.
When building with [cheribuild] this is usually handled automatically,
but if necessary the `--clean` flag can be used.  When building
directly, avoid the `-DNO_CLEAN` make option.

### Flag days on [master]

| Date       | Commit      | Description | Required action |
| ---------- | ----------- | --- | --- |

> | 2020-05-11 | [7e76d8f71] | C/C++ ABI changes | Update to [LLVM 8a3b2571b4] and a clean rebuild.|

Author: The LLVM hash here is wrong, I'll fix before commit.
